File: src/main/java/seremis/geninfusion/api/soul/EnumAlleleType.java
package seremis.geninfusion.api.soul;

import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import seremis.geninfusion.api.soul.util.ModelPart;

public enum EnumAlleleType {
    BOOLEAN(Boolean.class),
    BYTE(Byte.class),
    SHORT(Short.class),
    INTEGER(Integer.class),
    FLOAT(Float.class),
    DOUBLE(Double.class),
    LONG(Long.class),
    STRING(String.class),
    CLASS(Class.class),
    MODELPART(ModelPart.class),
    ITEMSTACK(ItemStack.class),
    BOOLEAN_ARRAY(boolean[].class),
    BYTE_ARRAY(byte[].class),
    SHORT_ARRAY(short[].class),
    INTEGER_ARRAY(int[].class),
    FLOAT_ARRAY(float[].class),
    DOUBLE_ARRAY(double[].class),
    LONG_ARRAY(long[].class),
    STRING_ARRAY(String[].class),
    CLASS_ARRAY(Class[].class),
    ITEMSTACK_ARRAY(ItemStack[].class),
    MODELPART_ARRAY(ModelPart[].class);

    public Class<?> clzz;

    EnumAlleleType(Class<?> clzz) {
        this.clzz = clzz;
    }

    public static EnumAlleleType forClass(Class<?> clzz) {
        for(EnumAlleleType type : values()) {
            if(type.clzz.equals(clzz)) {
                return type;
            }
        }
        return null;
    }

    public NBTTagCompound writeValueToNBT(NBTTagCompound compound, String name, Object value) {
        switch(this) {
            case BOOLEAN: compound.setBoolean(name, (Boolean) value); break;
            case BYTE: compound.setByte(name, (Byte) value); break;
            case SHORT: compound.setShort(name, (Short) value); break;
            case INTEGER: compound.setInteger(name, (Integer) value); break;
            case FLOAT: compound.setFloat(name, (Float) value); break;
            case DOUBLE: compound.setDouble(name, (Double) value); break;
            case LONG: compound.setLong(name, (Long) value); break;
            case STRING: compound.setString(name, (String) value); break;
            case CLASS: compound.setString(name, ((Class<?>) value).getName()); break;
            case MODELPART: compound.setTag(name, ((ModelPart) value).writeToNBT(new NBTTagCompound())); break;
            case ITEMSTACK: compound.setTag(name, ((ItemStack) value).writeToNBT(new NBTTagCompound())); break;
            case BOOLEAN_ARRAY: {
                for(int i = 0; i < ((boolean[]) value).length; i++) {
                    compound.setBoolean(name + "." + i, ((boolean[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((boolean[]) value).length);
                break;
            }
            case BYTE_ARRAY: {
                for(int i = 0; i < ((byte[]) value).length; i++) {
                    compound.setByte(name + "." + i, ((byte[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((byte[]) value).length);
                break;
            }
            case SHORT_ARRAY: {
                for(int i = 0; i < ((short[]) value).length; i++) {
                    compound.setShort(name + "." + i, ((short[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((short[]) value).length);
                break;
            }
            case INTEGER_ARRAY: {
                for(int i = 0; i < ((int[]) value).length; i++) {
                    compound.setInteger(name + "." + i, ((int[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((int[]) value).length);
                break;
            }
            case FLOAT_ARRAY: {
                for(int i = 0; i < ((float[]) value).length; i++) {
                    compound.setFloat(name + "." + i, ((float[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((float[]) value).length);
                break;
            }
            case DOUBLE_ARRAY: {
                for(int i = 0; i < ((double[]) value).length; i++) {
                    compound.setDouble(name + "." + i, ((double[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((double[]) value).length);
                break;
            }
            case LONG_ARRAY: {
                for(int i = 0; i < ((long[]) value).length; i++) {
                    compound.setLong(name + "." + i, ((long[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((long[]) value).length);
                break;
            }
            case STRING_ARRAY: {
                for(int i = 0; i < ((String[]) value).length; i++) {
                    compound.setString(name + "." + i, ((String[]) value)[i]);
                }
                compound.setInteger(name + ".length", ((String[]) value).length);
                break;
            }
            case CLASS_ARRAY: {
                for(int i = 0; i < ((Class<?>[]) value).length; i++) {
                    compound.setString(name + "." + i, ((Class<?>[]) value)[i].getName());
                }
                compound.setInteger(name + ".length", ((Class<?>[]) value).length);
                break;
            }
            case ITEMSTACK_ARRAY: {
                for(int i = 0; i < ((ItemStack[]) value).length; i++) {
                    compound.setTag(name + "." + i, ((ItemStack[]) value)[i].writeToNBT(new NBTTagCompound()));
                }
                compound.setInteger(name + ".length", ((ItemStack[]) value).length);
                break;
            }
            case MODELPART_ARRAY: {
                for(int i = 0; i < ((ModelPart[]) value).length; i++) {
                    compound.setTag(name + "." + i, ((ModelPart[]) value)[i].writeToNBT(new NBTTagCompound()));
                }
                compound.setInteger(name + ".length", ((ModelPart[]) value).length);
                break;
            }
        }
        return compound;
    }

    public Object readValueFromNBT(NBTTagCompound compound, String name) {
        Object result = null;

        switch(this) {
            case BOOLEAN: result = compound.getBoolean(name); break;
            case BYTE: result = compound.getByte(name); break;
            case SHORT: result = compound.getShort(name); break;
            case INTEGER: result = compound.getInteger(name); break;
            case FLOAT: result = compound.getFloat(name); break;
            case DOUBLE: result = compound.getDouble(name); break;
            case LONG: result = compound.getLong(name); break;
            case STRING: result = compound.getString(name); break;
            case CLASS: try{ result = Class.forName(compound.getString(name)); } catch(Exception e) {} break;
            case ITEMSTACK: ItemStack.loadItemStackFromNBT(compound.getCompoundTag(name)); break;
            case MODELPART: ModelPart.fromNBT(compound.getCompoundTag(name)); break;
            case BOOLEAN_ARRAY: {
                int length = compound.getInteger(name + ".length");
                boolean[] out = new boolean[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getBoolean(name + "." + i);
                }
                result = out;
                break;
            }
            case BYTE_ARRAY: {
                int length = compound.getInteger(name + ".length");
                byte[] out = new byte[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getByte(name + "." + i);
                }
                result = out;
                break;
            }
            case SHORT_ARRAY: {
                int length = compound.getInteger(name + ".length");
                short[] out = new short[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getShort(name + "." + i);
                }
                result = out;
                break;
            }
            case INTEGER_ARRAY: {
                int length = compound.getInteger(name + ".length");
                int[] out = new int[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getInteger(name + "." + i);
                }
                result = out;
                break;
            }
            case FLOAT_ARRAY: {
                int length = compound.getInteger(name + ".length");
                float[] out = new float[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getFloat(name + "." + i);
                }
                result = out;
                break;
            }
            case DOUBLE_ARRAY: {
                int length = compound.getInteger(name + ".length");
                double[] out = new double[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getDouble(name + "." + i);
                }
                result = out;
                break;
            }
            case LONG_ARRAY: {
                int length = compound.getInteger(name + ".length");
                long[] out = new long[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getLong(name + "." + i);
                }
                result = out;
                break;
            }
            case STRING_ARRAY: {
                int length = compound.getInteger(name + ".length");
                String[] out = new String[length];
                for(int i = 0; i < length; i++) {
                    out[i] = compound.getString(name + "." + i);
                }
                result = out;
                break;
            }
            case CLASS_ARRAY: {
                int length = compound.getInteger(name + ".length");
                Class<?>[] out = new Class<?>[length];
                try {
                    for(int i = 0; i < length; i++) {
                        out[i] = Class.forName(compound.getString(name + "." + i));
                    }
                } catch (Exception e) {}
                result = out;
                break;
            }
            case ITEMSTACK_ARRAY: {
                int length = compound.getInteger(name + ".length");
                ItemStack[] out = new ItemStack[length];
                for(int i = 0; i < length; i++) {
                    out[i] = ItemStack.loadItemStackFromNBT(compound.getCompoundTag(name + "." + i));
                }
                result = out;
                break;
            }
            case MODELPART_ARRAY: {
                int length = compound.getInteger(name + ".length");
                ModelPart[] out = new ModelPart[length];
                for(int i = 0; i < length; i++) {
                    out[i] = ModelPart.fromNBT(compound.getCompoundTag(name + "." + i));
                }
                result = out;
                break;
            }
        }
        return result;
    }
}

File: src/main/java/seremis/geninfusion/api/soul/EnumAlleleType.scala
package seremis.geninfusion.api.soul

import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import seremis.geninfusion.api.soul.util.ModelPart
//remove if not needed

object EnumAlleleType extends Enumeration {

    val BOOLEAN = new EnumAlleleType(classOf[Boolean])

    val BYTE = new EnumAlleleType(classOf[Byte])

    val SHORT = new EnumAlleleType(classOf[Short])

    val INTEGER = new EnumAlleleType(classOf[Integer])

    val FLOAT = new EnumAlleleType(classOf[Float])

    val DOUBLE = new EnumAlleleType(classOf[Double])

    val LONG = new EnumAlleleType(classOf[Long])

    val STRING = new EnumAlleleType(classOf[String])

    val CLASS = new EnumAlleleType(classOf[Class[_]])

    val MODELPART = new EnumAlleleType(classOf[ModelPart])

    val ITEMSTACK = new EnumAlleleType(classOf[ItemStack])

    val BOOLEAN_ARRAY = new EnumAlleleType(classOf[Array[Boolean]])

    val BYTE_ARRAY = new EnumAlleleType(classOf[Array[Byte]])

    val SHORT_ARRAY = new EnumAlleleType(classOf[Array[Short]])

    val INTEGER_ARRAY = new EnumAlleleType(classOf[Array[Int]])

    val FLOAT_ARRAY = new EnumAlleleType(classOf[Array[Float]])

    val DOUBLE_ARRAY = new EnumAlleleType(classOf[Array[Double]])

    val LONG_ARRAY = new EnumAlleleType(classOf[Array[Long]])

    val STRING_ARRAY = new EnumAlleleType(classOf[Array[String]])

    val CLASS_ARRAY = new EnumAlleleType(classOf[Array[Class[_]]])

    val ITEMSTACK_ARRAY = new EnumAlleleType(classOf[Array[ItemStack]])

    val MODELPART_ARRAY = new EnumAlleleType(classOf[Array[ModelPart]])

    class EnumAlleleType(var clzz: Class[_]) extends Val {

        def writeValueToNBT(compound: NBTTagCompound, name: String, value: AnyRef): NBTTagCompound = {
            this match {
                case BOOLEAN => compound.setBoolean(name, value.asInstanceOf[java.lang.Boolean])
                case BYTE => compound.setByte(name, value.asInstanceOf[java.lang.Byte])
                case SHORT => compound.setShort(name, value.asInstanceOf[java.lang.Short])
                case INTEGER => compound.setInteger(name, value.asInstanceOf[java.lang.Integer])
                case FLOAT => compound.setFloat(name, value.asInstanceOf[java.lang.Float])
                case DOUBLE => compound.setDouble(name, value.asInstanceOf[java.lang.Double])
                case LONG => compound.setLong(name, value.asInstanceOf[java.lang.Long])
                case STRING => compound.setString(name, value.asInstanceOf[String])
                case CLASS => compound.setString(name, value.asInstanceOf[Class[_]].getName)
                case MODELPART => compound.setTag(name, value.asInstanceOf[ModelPart].writeToNBT(new NBTTagCompound()))
                case ITEMSTACK => compound.setTag(name, value.asInstanceOf[ItemStack].writeToNBT(new NBTTagCompound()))
                case BOOLEAN_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Boolean]].length) {
                        compound.setBoolean(name + "." + i, value.asInstanceOf[Array[Boolean]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Boolean]].length)

                }
                case BYTE_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Byte]].length) {
                        compound.setByte(name + "." + i, value.asInstanceOf[Array[Byte]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Byte]].length)

                }
                case SHORT_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Short]].length) {
                        compound.setShort(name + "." + i, value.asInstanceOf[Array[Short]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Short]].length)

                }
                case INTEGER_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Int]].length) {
                        compound.setInteger(name + "." + i, value.asInstanceOf[Array[Int]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Int]].length)

                }
                case FLOAT_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Float]].length) {
                        compound.setFloat(name + "." + i, value.asInstanceOf[Array[Float]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Float]].length)

                }
                case DOUBLE_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Double]].length) {
                        compound.setDouble(name + "." + i, value.asInstanceOf[Array[Double]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Double]].length)

                }
                case LONG_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Long]].length) {
                        compound.setLong(name + "." + i, value.asInstanceOf[Array[Long]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Long]].length)

                }
                case STRING_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[String]].length) {
                        compound.setString(name + "." + i, value.asInstanceOf[Array[String]](i))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[String]].length)

                }
                case CLASS_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[Class[_]]].length) {
                        compound.setString(name + "." + i, value.asInstanceOf[Array[Class[_]]](i).getName)
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[Class[_]]].length)

                }
                case ITEMSTACK_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[ItemStack]].length) {
                        compound.setTag(name + "." + i, value.asInstanceOf[Array[ItemStack]](i).writeToNBT(new NBTTagCompound()))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[ItemStack]].length)

                }
                case MODELPART_ARRAY => {
                    for(i <- 0 until value.asInstanceOf[Array[ModelPart]].length) {
                        compound.setTag(name + "." + i, value.asInstanceOf[Array[ModelPart]](i).writeToNBT(new NBTTagCompound()))
                    }
                    compound.setInteger(name + ".length", value.asInstanceOf[Array[ModelPart]].length)
                }
            }
            compound
        }

        def readValueFromNBT(compound: NBTTagCompound, name: String): AnyRef = {
            var result: AnyRef = null
            this match {
                case BOOLEAN => result = compound.getBoolean(name)
                case BYTE => result = compound.getByte(name)
                case SHORT => result = compound.getShort(name)
                case INTEGER => result = compound.getInteger(name)
                case FLOAT => result = compound.getFloat(name)
                case DOUBLE => result = compound.getDouble(name)
                case LONG => result = compound.getLong(name)
                case STRING => result = compound.getString(name)
                case CLASS => try {
                    result = Class.forName(compound.getString(name))
                } catch {
                    case e: Exception =>
                }
                case ITEMSTACK => ItemStack.loadItemStackFromNBT(compound.getCompoundTag(name))
                case MODELPART => ModelPart.fromNBT(compound.getCompoundTag(name))
                case BOOLEAN_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Boolean](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getBoolean(name + "." + i)
                    }
                    result = out

                }
                case BYTE_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Byte](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getByte(name + "." + i)
                    }
                    result = out

                }
                case SHORT_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Short](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getShort(name + "." + i)
                    }
                    result = out

                }
                case INTEGER_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Int](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getInteger(name + "." + i)
                    }
                    result = out

                }
                case FLOAT_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Float](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getFloat(name + "." + i)
                    }
                    result = out

                }
                case DOUBLE_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Double](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getDouble(name + "." + i)
                    }
                    result = out

                }
                case LONG_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Long](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getLong(name + "." + i)
                    }
                    result = out

                }
                case STRING_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[String](length)
                    for (i <- 0 until length) {
                        out(i) = compound.getString(name + "." + i)
                    }
                    result = out

                }
                case CLASS_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[Class[_]](length)
                    try {
                        for (i <- 0 until length) {
                            out(i) = Class.forName(compound.getString(name + "." + i))
                        }
                    } catch {
                        case e: Exception =>
                    }
                    result = out
                }
                case ITEMSTACK_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[ItemStack](length)
                    for (i <- 0 until length) {
                        out(i) = ItemStack.loadItemStackFromNBT(compound.getCompoundTag(name + "." + i))
                    }
                    result = out
                }
                case MODELPART_ARRAY => {
                    val length = compound.getInteger(name + ".length")
                    val out = Array.ofDim[ModelPart](length)
                    for (i <- 0 until length) {
                        out(i) = ModelPart.fromNBT(compound.getCompoundTag(name + "." + i))
                    }
                    result = out
                }
            }
            result
        }
    }

    def forClass(clzz: Class[_]): EnumAlleleType = {
        values.find(_.clzz == clzz).getOrElse(null)
    }

    implicit def convertValue(v: Value): EnumAlleleType = v.asInstanceOf[EnumAlleleType]
}